core/modules/czar/Czar.cc

```diff
@@ -229,58 +227,102 @@ Czar::killQuery(std::string const& query, std::string const& clientId) {
         }
         uq = iter->second.lock();
+    } else if (ccontrol::UserQueryType::isCancel(query, queryId)) {
+        LOGS(_log, LOG_LVL_DEBUG, "query ID: " << queryId);
+        std::lock_guard<std::mutex> lock(_mutex);
+
+        // find it in the client map based in client/thread id
+        auto iter = _idToQuery.find(queryId);
+        if (iter == _idToQuery.end()) {
+            LOGS(_log, LOG_LVL_INFO, "Cannot find query id: " << queryId);
+            throw std::runtime_error("Unknown or finished query ID: " + query);
+        }
+        uq = iter->second.lock();
+    } else {
+        throw std::runtime_error("Failed to parse query: " + query);
     }
 
     // assume this cannot fail or throw
-    LOGS(_log, LOG_LVL_DEBUG, "Killing query for thread: " << threadId);
     if (uq) {
+        LOGS(_log, LOG_LVL_DEBUG, "Killing query: " << uq->getQueryId());
         // query killing can potentially take very long and we do now want to block
         // proxy from serving other requests so run it in a detached thread
         std::thread killThread([uq, threadId]() {
             uq->kill();
-            LOGS(_log, LOG_LVL_DEBUG, "Finished killing query for thread: " << threadId);
+            LOGS(_log, LOG_LVL_DEBUG, "Finished killing query: " << uq->getQueryId());
         });
         killThread.detach();
+    } else {
+        LOGS(_log, LOG_LVL_DEBUG, "Query has expired/finished: " << query);
+        throw std::runtime_error("Query has already finished: " + query);
     }
 }
 
 void
-Czar::_updateQueryHistory(std::string const& clientId,
-                          int threadId,
-                          ccontrol::UserQuery::Ptr const& uq) {
+Czar::_cleanupQueryHistoryLocked() {
+    // mutex must be locked
 
-    std::lock_guard<std::mutex> lock(_mutex);
-
-    // first cleanup client query map from completed queries
+    // first cleanup client query maps from completed queries
     for (auto iter = _clientToQuery.begin(); iter != _clientToQuery.end(); ) {
         if (iter->second.expired()) {
             iter = _clientToQuery.erase(iter);
         } else {
             ++ iter;
         }
     }
+    for (auto iter = _idToQuery.begin(); iter != _idToQuery.end(); ) {
+        if (iter->second.expired()) {
+            iter = _idToQuery.erase(iter);
+        } else {
+            ++ iter;
+        }
+    }
+}
+
+void
+Czar::_cleanupQueryHistory() {
+    std::lock_guard<std::mutex> lock(_mutex);
+    _cleanupQueryHistoryLocked();
+}
+
+void
+Czar::_updateQueryHistory(std::string const& clientId,
+                          int threadId,
+                          ccontrol::UserQuery::Ptr const& uq) {
+
+    std::lock_guard<std::mutex> lock(_mutex);
+
+    // first cleanup client query maps from completed queries
+    _cleanupQueryHistoryLocked();
 
     // remember query (weak pointer) in case we want to kill query
+    if (uq->getQueryId() != QueryId(0)) {
+        _idToQuery.insert(std::make_pair(uq->getQueryId(), uq));
```

Comment:
Would it be better to make the weak pointer for uq before the make_pair?

Reply:
In theory it does not matter, in practice I'd prefer to write Python-like:

    _idToQuery[uq->getQueryId()] = uq;

and not to worry about nonsensical conversions (but it's less efficient in C++). I think there is a reason why weak_ptr conversion constructor is implicit so it should be OK to use with this sort of pattern.